akka-typed/src/test/scala/akka/typed/ActorContextSpec.scala

```diff
@@ -0,0 +1,508 @@
+
+  case class Ping(replyTo: ActorRef[Pong]) extends Command
+  sealed trait Pong extends Event
+  case object Pong1 extends Pong
+  case object Pong2 extends Pong
+
+  case class Miss(replyTo: ActorRef[Missed.type]) extends Command
+  case object Missed extends Event
+
+  case class Renew(replyTo: ActorRef[Renewed.type]) extends Command
+  case object Renewed extends Event
+
+  case class Throw(ex: Exception) extends Command
+
+  case class MkChild(name: Option[String], monitor: ActorRef[GotSignal], replyTo: ActorRef[Created]) extends Command
+  case class Created(ref: ActorRef[Command]) extends Event
+
+  case class SetTimeout(duration: FiniteDuration, replyTo: ActorRef[TimeoutSet.type]) extends Command
+  case object TimeoutSet extends Event
+
+  case class Schedule[T](delay: FiniteDuration, target: ActorRef[T], msg: T, replyTo: ActorRef[Scheduled.type]) extends Command
+  case object Scheduled extends Event
+
+  case object Stop extends Command
+
+  case class Kill(name: String, replyTo: ActorRef[Killed.type]) extends Command
+  case object Killed extends Event
+
+  case class Watch(ref: ActorRef[Nothing], replyTo: ActorRef[Watched.type]) extends Command
+  case object Watched extends Event
+
+  case class Unwatch(ref: ActorRef[Nothing], replyTo: ActorRef[Unwatched.type]) extends Command
+  case object Unwatched extends Event
+
+  case class GetInfo(replyTo: ActorRef[Info]) extends Command
+  case class Info(self: ActorRef[Command], props: Props[Command], system: ActorSystem[Nothing]) extends Event
+
+  case class GetChild(name: String, replyTo: ActorRef[Child]) extends Command
+  case class Child(c: Option[ActorRef[Nothing]]) extends Event
+
+  case class GetChildren(replyTo: ActorRef[Children]) extends Command
+  case class Children(c: Set[ActorRef[Nothing]]) extends Event
+
+  case class ChildEvent(event: Event) extends Event
+
+  case class BecomeInert(replyTo: ActorRef[BecameInert.type]) extends Command
+  case object BecameInert extends Event
+
+  def subject(monitor: ActorRef[GotSignal]): Behavior[Command] =
+    FullTotal {
+      case Sig(ctx, signal) ⇒
+        monitor ! GotSignal(signal)
+        if (signal.isInstanceOf[Failed]) ctx.setFailureResponse(Failed.Restart)
+        Same
+      case Msg(ctx, message) ⇒ message match {
+        case Ping(replyTo) ⇒
+          replyTo ! Pong1
+          Same
+        case Miss(replyTo) ⇒
+          replyTo ! Missed
+          Unhandled
+        case Renew(replyTo) ⇒
+          replyTo ! Renewed
+          subject(monitor)
+        case Throw(ex) ⇒
+          throw ex
+        case MkChild(name, mon, replyTo) ⇒
+          val child = name match {
+            case None    ⇒ ctx.spawn(Props(subject(mon)))
+            case Some(n) ⇒ ctx.spawn(Props(subject(mon)), n)
+          }
+          replyTo ! Created(child)
+          Same
+        case SetTimeout(d, replyTo) ⇒
+          ctx.setReceiveTimeout(d)
+          replyTo ! TimeoutSet
+          Same
```

Comment:
Endre is right in that the context necessarily is impure, but not right in that it does not make sense to expose Actors that are exactly as pure as they can be :-) The Pure class that I have over in my development branch is not part of this PR because it is not ready yet. The idea is that invoking the PureBehavior will return effects to be performed and the next behavior—or more precisely the execution of the monadically represented effects results in the next behavior. It is obvious and unavoidable that this will have severe performance implications, therefore it cannot be the basic model that we expose.